File: Machine Learning/Deep Q/RLController.py
import keras
import random
import math
import numpy as np
from collections import deque
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam



class Agent:
    '''
    'Agent' class that represents the actions taken by a single neural network.
    '''

    def __init__(self, neural_net):
        '''
        Initialise class with the neural network.
        '''
        self.neural_net = neural_net
        self.memory = []
        self.performance = []

        self.epsilon = 1                # Exploration factor
        self.epsilon_decay =  0.995     # Less exploration with time
        self.epsilon_min =  0.01        # Still non-zero exploration after long time

        self.state = deque(maxlen=2)
        self.state.append(np.reshape([0, 0], [1, 2]))
        self.action = 4
    
    def perform_action(self, state: list, reward: float, done: bool):
        '''
        Takes state as [angle, velocity] and reward representing reward of previous action.
        Returns outputs as [leg_angle, torso_angle].
        '''
        self.state.append(np.reshape(state, [1, 2]))
        self.memory.append((self.state[0], self.action, reward, self.state[1], done))
        self.performance.append(reward)

        self.epsilon = max(self.epsilon_min, min(self.epsilon, 1.0 - math.log10((len(self.memory)) * self.epsilon_decay / 10000))) # Decay epsilon to reduce exploration
        if np.random.random() <= self.epsilon:
            self.action = random.randrange(0, 4)                     # Explore action space
        else:
            self.action = np.argmax(self.neural_net.predict(self.state[1]))  # Explore rewarding states
        
        if self.action == 0:        # Legs out
            output = [1, 0]
        elif self.action == 1:      # Legs in
            output = [-1, 0]
        elif self.action == 2:      # Torso out
            output = [0, 1]
        elif self.action == 3:      # Torso in
            output = [0, -1]
        else:                       # Do nothing
            output = [0, 0]
        
        return output

class Controller(list):
    '''
    'Controller' class that manages the updating of neural networks for a list of agents.
    '''

    def __init__(self):
        '''
        Initialise the class with learning parameters.
        '''
        self.alpha =  0.01               # Learning rate
        self.alpha_decay = 0.01          # Using previous actions to predict more with time
        self.gamma = 1                   # Discout Factor
    
    def make_agent(self):
        '''
        Create agent using Keras neural network.
        '''
        neural_net = Sequential()
        neural_net.add(Dense(52, input_dim=2, activation='tanh'))
        neural_net.add(Dense(128, activation='tanh'))
        neural_net.add(Dense(5, activation='linear'))
        neural_net.compile(loss='mse', optimizer=Adam(lr=self.alpha, decay=self.alpha_decay))
         
        super().append(Agent(neural_net))

    def replay(self, agent: Agent, batch_size: int = 64):
        '''
        Trains the model based on remembered information.
        '''
        x_batch, y_batch = [], []
        mini_batch = random.sample(agent.memory, min(len(agent.memory), batch_size))
    
        for state, action, reward, next_state, done in mini_batch:
            y_target = agent.neural_net.predict(state)
            if done:
                y_target[0][action] = reward
            else:
                y_target[0][action] = reward + self.gamma * np.max(agent.neural_net.predict(next_state)[0])
            x_batch.append(state[0])
            y_batch.append(y_target[0])
            
        agent.neural_net.fit(np.array(x_batch), np.array(y_batch), batch_size=len(x_batch), verbose=0)
    
    def step(self):
        '''
        Run RL algorithms based on memory of each agent.
        '''
        for agent in self:
            self.replay(agent)
        
        return self





# if __name__ == "__main__":
#     '''
#     Testing.
#     '''
#     c = Controller()
#     c.make_agent()

#     for i in range(20):
#         print(c[0].perform_action([1,i], i, False))
#         c.step()
    
#     print(c[0].perform_action([1,i], i, True))
#     c.step()
#     print(c[0].performance)

File: Machine Learning/Deep Q/RLController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Interfaces;
using Keras;

// input_size = 2
// output_size = 2

namespace Interfaces.Algorithms {
    class Agent : IEntity {
        var neural_net;
        int epsilon;
        int epsilon_decay;
        int epsilon_min;
        float[][] prev_states;
        int action;

        public Agent(var neural_net) {
            neural_net = neural_net;
            List<List<var>> memory = new List<List<int>>();
            List<int> performance = new List<int>();

            epsilon = 1;
            epsilon_decay = 0.995;
            epsilon_min = 0.01;

            prev_states = new float[2][2];
            prev_state[0][0] = 0
            prev_state[0][1] = 0
            action = 0;
        }

        public int[] perform_action(float[] state, float reward, bool done) {
            prev_states[0] = prev_states[1];
            prev_states[1] = state;

            memory.Add(new List<var>{prev_states[0], action, reward, prev_state[1], done})
            performance.Add(reward)

            epsilon = Math.Max(epsilon_min, Math.Min(epsilon, 1.0 - Math.Log10(memory.Count * epsilon_decay / 10000)));
            Random random = new Random();
            if (random.NextDouble() <= epsilon) {
                action = random.Next(0, 5)
            } else {
                prediction = neural_net.predict(state[1]);
                int maxValue = prediction.Max();
                int maxIndex = prediction.ToList().IndexOf(maxValue);
                action = maxIndex;
            }

            int[] output = new int[2]{0, 0};
            if (action == 0) {
                output[0] = 1;
            } else if (action == 1) {
                output[0] = -1;
            } else if (action == 2) {
                output[1] = 1;
            } else if (action == 3) {
                output[1] = -1;
            }

            return output;
        }
    }

    class Controller : IGenerationController {
        float alpha;
        float alpha_decay;
        float gamma;
        List<Agent> agents = new List<Agent>();

        public Controller() {
            alpha = 0.01;
            alpha_decay = 0.01;
            gamma = 1.0;
        }

        public void make_agent() {
            var neural_net = new Sequential();
            neural_net.Add(new Dense(52, activation: 'tanh', input_shape: new Shape(2)));
            neural_net.Add(new Dense(128, activation: 'tanh'));
            neural_net.Add(new Dense(2, activation: 'linear'));
            neural_net.Compile(loss: 'mse', optimizer: new Adam(lr: alpha, decay: alpha_decay));

            agents.Add(new Agent(neural_net));
        }

        public void replay(Agent agent, int batch_size: 64) {
            List<int> x_batch = new List<int>();
            List<int> y_batch = new List<int>();

            Random random = new Random();
            List<int> mini_batch = agent.memory.OrderBy(x => rnd.Next()).Take(Math.Min(agent.memory.Count, batch_size));

            for ((state, action, reward, next_state, done) in mini_batch) {
                int y_target = agent.neural_net.predict(state);
                if (done) {
                    y_target[0][action] = reward;
                } else {
                    y_target[0][action] = reward + gamma * Math.Max(agent.neural_net.predict(next_state)[0]);
                }
                x_batch.Add(state[0]);
                y_batch.Add(y_target[0]);
            }
            
            agent.neural_net.Fit(x_batch, y_batch, batch_size: x_batch.Count, verbose: 0);
        }

        public Controller step() {
            for (agent in agents) {
                this.replay(agent);
            }

            return this;
        }
    }
